File: dynamiclayout.erl
%% DynamicLayout Erlang API — generate layout JSON from Erlang code.
%%
%% Single-file module. Zero dependencies outside stdlib.
%% Works on OTP 24+.
%%
%% Example:
%%   B = dynamiclayout:begin("My Page"),
%%   B1 = dynamiclayout:fieldset(B, "Info"),
%%   B2 = dynamiclayout:label(B1, "Hello from Erlang!"),
%%   B3 = dynamiclayout:end_fieldset(B2),
%%   B4 = dynamiclayout:button(B3, "ok", "OK", "primary", true),
%%   io:format("~s~n", [dynamiclayout:end_layout(B4)]).

-module(dynamiclayout).
-export([begin/1, end_layout/1, fieldset/2, end_fieldset/1]).
-export([label/2, alert/3, badge/3, spacer/1, spacer/2]).
-export([input/3, input/4, textarea/3, textarea/4, checkbox/2, checkbox/3, rating/2]).
-export([button/4, button/5, section/2, section/3]).

begin(Title) ->
    {B, _} = #{buf => <<"{\"ui\":{">>, depth => 2, comma => false, cnt => 0},
    B1 = kv(B, <<"title">>, Title),
    arr(B1, <<"layout">>),
    B1.

end_layout(B) ->
    B1 = close_arr(B),
    B2 = B1#{comma := false},
    B3 = putc(B2, $,),
    B4 = puts(B3, <<"\"translations\":{},">>),
    B5 = puts(B4, <<"\"userAccess\":{\"cancel\":true}">>),
    puts(B5, <<"}}">>),
    binary_to_list(B5).

%% -- internal --
putc(B = #{buf := Buf}, C) -> B#{buf := <<Buf/binary, C>>}.
puts(B = #{buf := Buf}, S) -> B#{buf := <<Buf/binary, S/binary>>}.
indent(B = #{depth := D}) ->
    B1 = putc(B, $\n),
    lists:foldl(fun(_, Acc) -> puts(Acc, <<"  ">>) end, B1, lists:seq(1, D)).
cm(B = #{comma := true}) -> putc(B#{comma := false}, $,);
cm(B) -> B.

esc(B, S) ->
    B1 = putc(B, $"),
    B2 = lists:foldl(fun
        ($",  Acc) -> puts(Acc, <<"\\\"">>);
        ($\\, Acc) -> puts(Acc, <<"\\\\">>);
        ($\n, Acc) -> puts(Acc, <<"\\n">>);
        ($\r, Acc) -> puts(Acc, <<"\\r">>);
        ($\t, Acc) -> puts(Acc, <<"\\t">>);
        (C,   Acc) -> putc(Acc, C)
    end, B1, binary_to_list(S)),
    putc(B2, $").

kv(B, K, V) ->
    B1 = cm(B), B2 = putc(B1, $,), B3 = indent(B2),
    B4 = esc(B3, K), B5 = puts(B4, <<": ">>), B6 = esc(B5, V),
    B6#{comma => true}.

kvint(B, K, V) ->
    B1 = cm(B), B2 = putc(B1, $,), B3 = indent(B2),
    B4 = esc(B3, K), B5 = puts(B4, <<": ">>),
    putc(B5#{comma => true}, integer_to_list(V)).

kvbool(B, K) ->
    B1 = cm(B), B2 = putc(B1, $,), B3 = indent(B2),
    B4 = esc(B3, K), puts(B4#{comma => true}, <<": true">>).

obj(B = #{depth := D}) ->
    B1 = cm(B), B2 = putc(B1, $,), B3 = indent(B2),
    putc(B3#{depth := D + 1, comma := false}, ${).

close_obj(B = #{depth := D}) ->
    B1 = B#{depth := D - 1, comma := false}, B2 = indent(B1),
    putc(B2#{comma := true}, $}).

arr(B = #{depth := D}, K) ->
    B1 = cm(B), B2 = putc(B1, $,), B3 = indent(B2),
    B4 = esc(B3, K), B5 = puts(B4, <<": [">>),
    B5#{depth := D + 1, comma := false}.

close_arr(B = #{depth := D}) ->
    B1 = B#{depth := D - 1, comma := false}, B2 = indent(B1),
    putc(B2#{comma := true}, $]).

key(B = #{cnt := C}, P) ->
    {B#{cnt := C + 1}, list_to_binary([P, "_", integer_to_list(C + 1)])}.

%% -- public --
fieldset(B, Title) ->
    {B1, K} = key(B, <<"fs">>),
    B2 = obj(B1), B3 = kv(B2, <<"type">>, <<"FIELDSET">>),
    B4 = kv(B3, <<"key">>, K), B5 = kv(B4, <<"title">>, Title),
    arr(B5, <<"content">>).

end_fieldset(B) -> close_obj(close_arr(B)).

label(B, Text) ->
    {B1, K} = key(B, <<"l">>),
    B2 = obj(B1), B3 = kv(B2, <<"type">>, <<"LABEL">>),
    B4 = kv(B3, <<"key">>, K), B5 = kv(B4, <<"label">>, Text),
    close_obj(B5).

alert(B, Msg, Color) ->
    {B1, K} = key(B, <<"a">>),
    B2 = obj(B1), B3 = kv(B2, <<"type">>, <<"ALERT">>),
    B4 = kv(B3, <<"key">>, K), B5 = kv(B4, <<"message">>, Msg),
    close_obj(kv(B5, <<"color">>, list_to_binary(Color))).

badge(B, Title, Color) ->
    {B1, K} = key(B, <<"bd">>),
    B2 = obj(B1), B3 = kv(B2, <<"type">>, <<"BADGE">>),
    B4 = kv(B3, <<"key">>, K), B5 = kv(B4, <<"title">>, Title),
    close_obj(kv(B5, <<"color">>, list_to_binary(Color))).

spacer(B) -> spacer(B, 20).
spacer(B, Px) ->
    {B1, K} = key(B, <<"sp">>),
    B2 = obj(B1), B3 = kv(B2, <<"type">>, <<"SPACER">>),
    close_obj(kvint(kv(B3, <<"key">>, K), <<"width">>, Px)).

input(B, Id, Label) -> input(B, Id, Label, false).
input(B, Id, Label, Req) ->
    {B1, K} = key(B, <<"i">>),
    B2 = obj(B1), B3 = kv(B2, <<"type">>, <<"INPUT">>),
    B4 = kv(B3, <<"key">>, K), B5 = kv(B4, <<"id">>, Id),
    B6 = kv(B5, <<"label">>, Label),
    close_obj(case Req of true -> kvbool(B6, <<"required">>); _ -> B6 end).

checkbox(B, Id) -> checkbox(B, Id, <<"">>).
checkbox(B, Id, Label) ->
    {B1, K} = key(B, <<"cb">>),
    B2 = obj(B1), B3 = kv(B2, <<"type">>, <<"CHECKBOX">>),
    close_obj(kv(kv(kv(B3, <<"key">>, K), <<"id">>, Id), <<"label">>, Label)).

textarea(B, Id, Label) -> textarea(B, Id, Label, 3).
textarea(B, Id, Label, Rows) ->
    {B1, K} = key(B, <<"ta">>),
    B2 = obj(B1), B3 = kv(B2, <<"type">>, <<"TEXTAREA">>),
    B4 = kv(B3, <<"key">>, K), B5 = kv(B4, <<"id">>, Id),
    close_obj(kvint(kv(B5, <<"label">>, Label), <<"rows">>, Rows)).

rating(B, Id) ->
    {B1, K} = key(B, <<"rt">>),
    B2 = obj(B1), B3 = kv(B2, <<"type">>, <<"RATING">>),
    close_obj(kv(kv(B3, <<"key">>, K), <<"id">>, Id)).

button(B, Id, Title, Color) -> button(B, Id, Title, Color, false).
button(B, Id, Title, Color, Def) ->
    {B1, K} = key(B, <<"btn">>),
    B2 = obj(B1), B3 = kv(B2, <<"type">>, <<"BUTTON">>),
    B4 = kv(B3, <<"key">>, K), B5 = kv(B4, <<"id">>, Id),
    B6 = kv(B5, <<"title">>, Title),
    B7 = kv(B6, <<"color">>, list_to_binary(Color)),
    close_obj(case Def of true -> kvbool(B7, <<"default">>); _ -> B7 end).

section(B, Title) -> section(B, Title, <<"">>).
section(B, Title, Text) ->
    B1 = fieldset(B, Title),
    B2 = label(B1, Text),
    end_fieldset(B2).

File: dynamiclayout.exs
# DynamicLayout Elixir API — generate layout JSON from Elixir code.
#
# Single-file module. Zero dependencies outside stdlib.
# Works on Elixir 1.14+.
#
# Example:
#   alias DynamicLayout, as: DL
#   DL.begin("My Page")
#   |> DL.fieldset("Info")
#   |> DL.label("Hello from Elixir!")
#   |> DL.end_fieldset()
#   |> DL.button("ok", "OK", "primary", true)
#   |> DL.end_layout()
#   |> IO.puts()

defmodule DynamicLayout do
  defstruct buf: "", depth: 0, comma: false, cnt: 0

  def begin(title) do
    %__MODULE__{}
    |> puts(~s({"ui":{))
    |> struct!(depth: 2, comma: false)
    |> kv("title", title)
    |> arr("layout")
  end

  # internal
  defp putc(b, c), do: %{b | buf: b.buf <> <<c>>}
  defp puts(b, s), do: %{b | buf: b.buf <> s}
  defp indent(b), do: b |> putc(?\n) |> puts(String.duplicate("  ", b.depth))
  defp cm(b), do: if(b.comma, do: putc(%{b | comma: false}, ?,), else: b)
  defp esc(b, s) do
    b |> putc(?") |> puts(s |> String.replace("\\", "\\\\") |> String.replace("\"", "\\\"") |> String.replace("\n", "\\n")) |> putc(?")
  end
  defp kv(b, k, v) do
    b |> cm() |> putc(?,) |> indent() |> esc(k) |> puts(": ") |> esc(v) |> Map.put(:comma, true)
  end
  defp kvint(b, k, v) do
    b |> cm() |> putc(?,) |> indent() |> esc(k) |> puts(": #{v}") |> Map.put(:comma, true)
  end
  defp kvbool(b, k) do
    b |> cm() |> putc(?,) |> indent() |> esc(k) |> puts(": true") |> Map.put(:comma, true)
  end
  defp obj(b), do: b |> cm() |> putc(?,) |> indent() |> putc(?{) |> Map.merge(%{depth: b.depth + 1, comma: false})
  defp close_obj(b), do: %{b | depth: b.depth - 1} |> indent() |> putc(?}) |> Map.put(:comma, true)
  defp arr(b, k), do: b |> cm() |> putc(?,) |> indent() |> esc(k) |> puts(": [") |> Map.merge(%{depth: b.depth + 1, comma: false})
  defp close_arr(b), do: %{b | depth: b.depth - 1} |> indent() |> putc(?]) |> Map.put(:comma, true)
  defp key(b, p), do: {%{b | cnt: b.cnt + 1}, "#{p}_#{b.cnt + 1}"}

  def end_layout(b) do
    b |> close_arr() |> Map.put(:comma, false) |> putc(?,) |> puts(~s("translations":{},)) |> puts(~s("userAccess":{"cancel":true})) |> puts("}}") |> Map.get(:buf)
  end

  def fieldset(b, title) do
    {b, k} = key(b, "fs")
    b |> obj() |> kv("type", "FIELDSET") |> kv("key", k) |> kv("title", title) |> arr("content")
  end
  def end_fieldset(b), do: b |> close_arr() |> close_obj()

  def label(b, text) do
    {b, k} = key(b, "l")
    b |> obj() |> kv("type", "LABEL") |> kv("key", k) |> kv("label", text) |> close_obj()
  end
  def alert(b, msg, color \\ "info") do
    {b, k} = key(b, "a")
    b |> obj() |> kv("type", "ALERT") |> kv("key", k) |> kv("message", msg) |> kv("color", color) |> close_obj()
  end
  def badge(b, title, color \\ "primary") do
    {b, k} = key(b, "bd")
    b |> obj() |> kv("type", "BADGE") |> kv("key", k) |> kv("title", title) |> kv("color", color) |> close_obj()
  end
  def spacer(b, px \\ 20) do
    {b, k} = key(b, "sp")
    b |> obj() |> kv("type", "SPACER") |> kv("key", k) |> kvint("width", px) |> close_obj()
  end
  def input(b, id, label, required \\ false) do
    {b, k} = key(b, "i")
    b = b |> obj() |> kv("type", "INPUT") |> kv("key", k) |> kv("id", id) |> kv("label", label)
    b = if required, do: kvbool(b, "required"), else: b
    close_obj(b)
  end
  def checkbox(b, id, label \\ "") do
    {b, k} = key(b, "cb")
    b |> obj() |> kv("type", "CHECKBOX") |> kv("key", k) |> kv("id", id) |> kv("label", label) |> close_obj()
  end
  def textarea(b, id, label, rows \\ 3) do
    {b, k} = key(b, "ta")
    b = b |> obj() |> kv("type", "TEXTAREA") |> kv("key", k) |> kv("id", id) |> kv("label", label)
    b = if rows > 0, do: kvint(b, "rows", rows), else: b
    close_obj(b)
  end
  def rating(b, id, label \\ "") do
    {b, k} = key(b, "rt")
    b |> obj() |> kv("type", "RATING") |> kv("key", k) |> kv("id", id) |> kv("label", label) |> close_obj()
  end
  def button(b, id, title, color \\ "primary", is_default \\ false) do
    {b, k} = key(b, "btn")
    b = b |> obj() |> kv("type", "BUTTON") |> kv("key", k) |> kv("id", id) |> kv("title", title) |> kv("color", color)
    b = if is_default, do: kvbool(b, "default"), else: b
    close_obj(b)
  end
  def section(b, title, text) do
    b |> fieldset(title) |> label(text) |> end_fieldset()
  end
end
